src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,448 @@
+            {
+                var charType = model.Compilation.GetSpecialType(SpecialType.System_Char);
+                if (!IsExchangable(semanticFact, charType, foreachType, model.Compilation))
+                {
+                    return;
+                }
+
+                collectionNameSuggestion = "str";
+                explicitCastInterface = null;
+                countName = Length;
+                return;
+            }
+
+            // check ImmutableArray case 
+            if (collectionType.OriginalDefinition.Equals(model.Compilation.GetTypeByMetadataName(typeof(ImmutableArray<>).FullName)))
+            {
+                var indexer = GetInterfaceMember(collectionType, get_Item);
+                if (indexer != null)
+                {
+                    if (!IsExchangable(semanticFact, indexer.ReturnType, foreachType, model.Compilation))
+                    {
+                        return;
+                    }
+
+                    collectionNameSuggestion = "array";
+                    explicitCastInterface = null;
+                    countName = Length;
+                    return;
+                }
+            }
+
+            // go through all known interfaces we support next.
+            var knownCollectionInterfaces = s_KnownInterfaceNames.Select(
+                s => model.Compilation.GetTypeByMetadataName(s)).Where(t => !IsNullOrErrorType(t));
+
+            // for all interfaces, we suggest collection name as "list"
+            collectionNameSuggestion = "list";
+
+            // check type itself is interface case
+            if (collectionType.TypeKind == TypeKind.Interface && knownCollectionInterfaces.Contains(collectionType.OriginalDefinition))
+            {
+                var indexer = GetInterfaceMember(collectionType, get_Item);
+                if (indexer != null &&
+                    IsExchangable(semanticFact, indexer.ReturnType, foreachType, model.Compilation))
+                {
+                    explicitCastInterface = null;
+                    countName = Count;
+                    return;
+                }
+            }
+
+            // check regular cases (implicitly implemented)
+            ITypeSymbol explicitInterface = null;
+            foreach (var current in collectionType.AllInterfaces)
+            {
+                if (!knownCollectionInterfaces.Contains(current.OriginalDefinition))
+                {
+                    continue;
+                }
+
+                // see how the type implements the interface
+                var countSymbol = GetInterfaceMember(current, get_Count);
+                var indexerSymbol = GetInterfaceMember(current, get_Item);
+                if (countSymbol == null || indexerSymbol == null)
+                {
+                    continue;
+                }
+
+                var countImpl = collectionType.FindImplementationForInterfaceMember(countSymbol) as IMethodSymbol;
```

Comment:
We should probably share the code here with for-to-foreach refactoring?

Reply:
I am looking for specific interface, and he doesn't actually care specific interface. that's where this difference came.

Reply:
suggestion is tracked here - https://github.com/dotnet/roslyn/issues/25667

> let people convert to for if the type implements indexer and either Length or Count
#25460 (review)